docker/complement/Dockerfile

```diff
@@ -1,22 +1,40 @@
-# A dockerfile which builds an image suitable for testing Synapse under
-# complement.
+# This dockerfile builds on top of 'docker/Dockerfile-worker' in matrix-org/synapse
+# by including a built-in postgres instance, as well as setting up the homeserver so
+# that it is ready for testing via Complement.
+#
+# Instructions for building this image from those it depends on is detailed in this guide:
+# https://github.com/matrix-org/synapse/blob/develop/docker/README-testing.md#testing-with-postgresql-and-single-or-multi-process-synapse
+FROM matrixdotorg/synapse-workers
+
+# Install postgresql
+RUN apt-get update && \
+  DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y postgresql-13
+
+# Configure a user and create a database for Synapse
+RUN pg_ctlcluster 13 main start &&  su postgres -c "echo \
+ \"ALTER USER postgres PASSWORD 'somesecret'; \
+ CREATE DATABASE synapse \
+  ENCODING 'UTF8' \
+  LC_COLLATE='C' \
+  LC_CTYPE='C' \
+  template=template0;\" | psql" && pg_ctlcluster 13 main stop
+
+# Modify the shared homeserver config with postgres support, certificate setup
+# and the disabling of rate-limiting
+COPY conf-workers/workers-shared.yaml /conf/workers/shared.yaml
 
-ARG SYNAPSE_VERSION=latest
-
-FROM matrixdotorg/synapse:${SYNAPSE_VERSION}
-
-ENV SERVER_NAME=localhost
-
-COPY conf/* /conf/
+WORKDIR /data
 
-# generate a signing key
-RUN generate_signing_key -o /conf/server.signing.key
```

Comment:
This isn't something I've touched (although the overall diff will make that less than obvious) — presumably it generated a signing key before. That said, this one eluded me for a bit!

It's in `start.py` (called by `configure_workers_and_start.py`):
https://github.com/matrix-org/synapse/blob/7f92ac4c1cbd7fae815c02b4920eb02ddf9458cb/docker/start.py#L115-L126

Reply:
I didn't write that file — it was already here: [`e4137c1` (#12881)](https://github.com/matrix-org/synapse/pull/12881/commits/e4137c13618f57f1f69c442644d287e6acc8b6dc#diff-8aa6e0870e6d530b9c43efd5affb47404f818d4136bb7de2459d8636d3e4485f); the diff is just confusing because I deleted the monolith-only image which is what the red part of this diff (that your comment is on) corresponds to